File: packages/rune-keyboard/ios/RuneKeyboardBridge.swift
//
//  RuneKeyboardBridge.swift
//  RuneKeyboard
//
//  Bridge module that exposes keyboard functionality to JavaScript
//  via the __modules.call() mechanism.
//

import Foundation
import RuneKit

final class RuneKeyboardBridge: NSObject, RuneModule {

  let name = "RuneKeyboard"
  private weak var keyboardModule: RuneKeyboardModule?

  init(keyboardModule: RuneKeyboardModule) {
    self.keyboardModule = keyboardModule
    super.init()
  }

  func call(method: String, args: Any?) throws -> Any? {
    print("[RuneKeyboardBridge] call(\(method))")
    switch method {
    case "dismiss":
      return handleDismiss()
    case "getState":
      return handleGetState()
    default:
      return ["error": "unsupported_method", "message": method]
    }
  }

  private func handleDismiss() -> [String: Any] {
    print("[RuneKeyboardBridge] handleDismiss() invoked from JS")
    keyboardModule?.dismissKeyboard()
    return ["success": true]
  }

  private func handleGetState() -> [String: Any] {
    return ["success": true]
  }
}

File: packages/rune-keyboard/android/RuneKeyboard/src/main/java/dev/rune/keyboard/RuneKeyboardBridge.kt
package dev.rune.keyboard

import com.rune.kit.runtime.RuneModule
import org.json.JSONObject

/**
 * Bridge module that exposes keyboard functionality to JavaScript
 * via the __modules.call() mechanism.
 */
class RuneKeyboardBridge(
    private val keyboardModule: RuneKeyboardModule
) : RuneModule {
    
    override val name: String = "RuneKeyboard"

    override val constants: Map<String, Any>?
        get() = keyboardModule.getInitialState().toMap()

    override fun call(method: String, args: Array<Any?>): JSONObject {
        return when (method) {
            "dismiss" -> handleDismiss()
            "getState" -> handleGetState()
            else -> errorResponse("unsupported_method", method)
        }
    }

    private fun handleDismiss(): JSONObject {
        keyboardModule.dismissKeyboard()
        return successResponse()
    }

    private fun handleGetState(): JSONObject {
        // Could be implemented to return current state if needed
        return successResponse()
    }

    private fun successResponse(): JSONObject {
        return JSONObject().apply {
            put("success", true)
        }
    }

    private fun errorResponse(error: String, message: String): JSONObject {
        return JSONObject().apply {
            put("error", error)
            put("message", message)
        }
    }
}
